### ISSUE TITLE
gRPC-Web Support

### ISSUE THREAD

**Issue by otto-dev** [akka/akka-grpc#695]
### Short description

Let's add support for [gRPC-Web](https://github.com/grpc/grpc-web)

### Details

Maybe we can add a configuration or alternative implementation that is compatible with the specifications of [gRPC-Web](https://github.com/grpc/grpc-web), so that browsers can access the Akka HTTP endpoints without another HTTP proxy.

[The specification doesn't look too complicated](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md#protocol-differences-vs-grpc-over-http2). I hacked around with it a bit today, and it didn't take too much to be able to communicate via unary calls.

All I had to do for unary calls to work was:
- base64 encode request and response bodies
- set headers

Streaming seems to be a bit more complicated. But probably quite doable for someone familiar with the gRPC protocol. From the documentation, and what I saw seems to be some extra metadata appended to the response bodies. From memory it was something like this:
```
3grpc-status: 0
Date: <curent date/time>
```
The first bytes (`3`) appear to be the length of the following body in binary format. It talks in the specs about these key-value pairs in the response body. I didn't play around with that part any more, mainly because I don't know enough about the `gRPC` specs to know what's going on.

I will attach some very hacky wrapper that made it possible for me to access a gRPC service from the browser - except streaming. See the following comment.

**Comment by otto-dev** [comment]
```scala
    // this is our gRPC service
    val service = SomeHandler(new SomeService())

    // we wrap our service to do some hacking.
    val handler: HttpRequest => Future[HttpResponse] = { request: HttpRequest =>
      val withoutEncoding = request.copy(
        // gRPC web doesn't play with gzip encoding. This is a hack to stop Akka HTTP from encoding the response
        headers = request.headers.filterNot(_.name == "grpc-accept-encoding"),

        // base64 body instead of plain bytes
        entity = request.entity.transformDataBytes(Flow[ByteString].map { v => ByteString(Base64.getDecoder.decode(v.toByteBuffer)) })
      )
      // now get the response of our service and modify it - then return it
      service(withoutEncoding).map { response =>
        // We need some special headers for gRPC web
        response.withHeaders(
          `Access-Control-Allow-Origin`.*,
          `Access-Control-Allow-Headers`("x-user-agent", "x-grpc-web", "content-type"),
          `Access-Control-Expose-Headers`("grpc-status", "grpc-message"),
        )
        // we need this content type
        val contentType = ContentType(MediaType.custom("application/grpc-web-text", binary = false), () => HttpCharsets.`UTF-8`)
        // base64 body instead of plain bytes for response
        val byteSource = response.entity.dataBytes.map(x => ByteString(Base64.getEncoder.encode(x.toByteBuffer)))
        val e: ResponseEntity = HttpEntity.CloseDelimited(contentType, byteSource)
        response.withEntity(e)
      }
    }
    val binding = Http().bindAndHandleAsync(
      service,
      interface = "127.0.0.1",
      port = 4000,
      connectionContext = HttpConnectionContext(),
    )
    binding.foreach { v => println(s"gRPC server bound to: ${v.localAddress}") }
    binding
```

**Comment by johanandren** [comment]
Possibly related: #654

**Comment by markns** [comment]
Just to note, in case anyone is doubting this (as I was) - akka-grpc supports grpc-web when the Java server is proxied by envoy (standard grpc-web setup).

**Comment by ennru** [comment]
This is implemented in #829 and will be available with 0.8.0.
